aiida/orm/implementation/general/calculation/job/__init__.py

```diff
@@ -1960,8 +1960,8 @@ def _presubmit(self, folder, use_unstored_links=False):
         folder.create_file_from_filelike(StringIO(script_content), script_filename)
 
         subfolder = folder.get_subfolder('.aiida', create=True)
-        subfolder.create_file_from_filelike(StringIO(json.dumps(job_tmpl)), 'job_tmpl.json')
-        subfolder.create_file_from_filelike(StringIO(json.dumps(calcinfo)), 'calcinfo.json')
+        subfolder.create_file_from_filelike(StringIO(six.text_type(json.dumps(job_tmpl)), 'job_tmpl.json'))
+        subfolder.create_file_from_filelike(StringIO(six.text_type(json.dumps(calcinfo)), 'calcinfo.json'))
```

Comment:
this is not the same, StringIO gives you a file-like object while six.text_type converts a str to a unicode on Python 2 and is a no-op on Python 3

Reply:
Hi, I disagree on this one. 

I know that the code looks ugly, but it’s to satisfy compatibility with Python 2 and 3. 
The underlying issue is really create_file_from_filelike() which uses shutil.copyfileobj internally. 
Minimal code that shows the difficult behaviour is below.

    # encoding: utf-8
    import io
    import json
    import shutil
    from six import StringIO as StringIO

    d_ascii={'a':'hello'}
    filelike = StringIO(json.dumps(d_ascii))

    with io.open('/tmp/test.dat', 'wb')  as dest_file:
        shutil.copyfileobj(filelike, dest_file)

This will work in Python2 but fail in Python3.
The reason is that the filelike object returned by StringIO is of the string type (due to coming from json.dumps). In Python 3, this is a unicode string which can’t be buffered in ‘wb’ mode,  which requires bytes.

One could change the io.open line to encode  to UTF-8 (and this is what I now do):
    with io.open('/tmp/test.dat', 'w’, encoding=’utf8’)  as dest_file:
but now Python 2 will fail as io.open demands a unicode object, but we are supplying a byte-string. 

A solution which satisfies both Python2 and Python3, while ensuring that the written file is UTF-8 encoded, is to make sure that what comes out of json.dumps is converted to a unicode string. 
Six.text_type() does this. At this point we haven’t encoded, but we eventually do this with io.open, and besides, json does encoding to UTF-8 internally. 

In the end, the following examples satisfies both Python 2 and 3:
    # encoding: utf-8
    import io
    import json
    import shutil
    
    import six
    from six import StringIO as StringIO

    d_ascii={'a':'hello'}
    filelike = StringIO(six.text_type(json.dumps(d_ascii)))
    with io.open('/tmp/test.dat', 'w', encoding='utf8')  as dest_file:
        shutil.copyfileobj(filelike, dest_file)

    d_unicode={'b': u'γεια σας'}
    filelike = StringIO(six.text_type(json.dumps(d_unicode)))
    with io.open('/tmp/test.dat', 'w', encoding='utf8')  as dest_file:
        shutil.copyfileobj(filelike, dest_file)

Reply:
The problem with using `six.text_type` to work around the inconsistent json API is that if you do not specify anything, it uses the default system encoding for decoding a `bytes` or a `str` into a unicode string, which is exactly the ambiguity we try to avoid in the first place.
But if you are adding the `encoding="utf8"`, then `six.text_type("", encoding="utf8")` runs on Python 2 and fails on Python 3.

I would rather prefer an explicit `if six.PY2: ...` switch or a (maintained and consistent) library like `simplejson`.

The other drawback of these kind of solutions is that in a year or two nobody will have a clue anymore why this wrapping is there, whether it is still needed or not and it will have spread to other places because people are doing copy&paste, especially when it comes to solving problems they don't know about.

Reply:
This is a fair point. 
I think the tidy thing to do, if we want to use a switch, is to have a json utility function, although I don't know where best to put this. Then it would be just a case of changing the json dump function calls where they occur.